### ISSUE TITLE
Improve error message E0608 when the type supports Index but Idx is missing trait bounds

### ISSUE THREAD

**Issue by Zannick** [rust-lang/rust#110373]
I tried this code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=3ea8873134cbec810fd7073a1e277ec2)):

```rust
use std::collections::HashMap;
use std::hash::Hash;

pub struct Edge<E> {
    id: E,
    src: usize,
    dst: usize,
    wt: u64
}

pub struct Graph<V, E> {
    nodes: Vec<V>,
    node_index_map: HashMap<V, usize>,
    edges: Vec<Edge<E>>,
}

pub fn build_graph<V, E>(v: V, e: E) -> Graph<V, E>
where V: Copy + Eq + Hash, E: Eq + Hash {
    let mut node_index_map = HashMap::new();
    let nodes = vec![v];
    node_index_map.insert(v, 0);
    let edges = vec![Edge { id: e, src: node_index_map[&v], dst: 0, wt: 1}];
    Graph {
        nodes,
        node_index_map,
        edges,
    }
}

impl<V, E> Graph<V, E> {
    pub fn node_index(&self, node: V) -> usize {
        self.node_index_map[&node]  // produces an error
    }
}
```

I expected to see this happen: rustc outputs an error saying that I need to add the `Eq` and `std::hash::Hash` traits on `V`.

Instead, this happened:

> error[[E0608]](https://doc.rust-lang.org/nightly/error_codes/E0608.html): cannot index into a value of type `HashMap<V, usize>`

which is jarring to see, and unhelpful, because HashMap does support Index, but only with particular trait bounds on `V`.

**Comment by Ezrashaw** [comment]
The reduced case:
```rust
use core::ops::Index;

trait Trait {}
struct Container;

impl<T: Trait> Index<T> for Container {
    type Output = T;

    fn index(&self, idx: T) -> &Self::Output {
        &idx
    }
}

fn main() {
    struct Containee;

    Container[Containee];
}
```
works as expected, perhaps it is caused when the `Index` implementer is not from the local crate?
